From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ruViolence <78062896+ruViolence@users.noreply.github.com>
Date: Wed, 15 Feb 2023 18:37:46 +0500
Subject: [PATCH] feat: configurable portal exit chunks generation

Allows to disable chunk generation when searching for an exit nether portal. Since the chunks in that scenario are generated synchronously, this saves a lot of performance when exiting to ungenerated area.

diff --git a/src/main/java/me/core/lamius/Lamius/LamiusConfig.java b/src/main/java/me/core/lamius/Lamius/LamiusConfig.java
index 97d95ad1b736ca9b61204c36c55a5f32377f13b6..d2d473a50e114d5283f5ba5b71f675488ad36161 100644
--- a/src/main/java/me/core/lamius/Lamius/LamiusConfig.java
+++ b/src/main/java/me/core/lamius/Lamius/LamiusConfig.java
@@ -290,6 +290,11 @@ public class LamiusConfig {
         dropUsePacketsInInventory = getBoolean("drop-use-packets-in-inventory", false);
     }
 
+    public static boolean generatePortalExitChunks;
+    private static void generatePortalExitChunks() {
+        generatePortalExitChunks = getBoolean("generate-portal-exit-chunks", false);
+    }
+
     // Paper start - packet limiter
     public static String packetLimiterKickMessage;
     public static PacketLimit packetLimiterAllPackets;
diff --git a/src/main/java/net/minecraft/server/PortalTravelAgent.java b/src/main/java/net/minecraft/server/PortalTravelAgent.java
index 897a92d1c193c3b33a679fbd4cded878710e1ed7..18124f264ca7fa74df1ea9376abe6614cf16dc14 100644
--- a/src/main/java/net/minecraft/server/PortalTravelAgent.java
+++ b/src/main/java/net/minecraft/server/PortalTravelAgent.java
@@ -154,10 +154,21 @@ public class PortalTravelAgent {
         } else {
             BlockPosition blockposition = new BlockPosition(x, y, z); // CraftBukkit
 
+            java.util.Set<ChunkCoordIntPair> notGeneratedChunks = me.core.lamius.Lamius.LamiusConfig.generatePortalExitChunks ? null : new HashSet<>(); // Lamius
             for (int l = -radius; l <= radius; ++l) {
                 BlockPosition blockposition1;
 
                 for (int i1 = -radius; i1 <= radius; ++i1) {
+                    // Lamius start
+                    if (notGeneratedChunks != null) {
+                        ChunkCoordIntPair chunkCoord = new ChunkCoordIntPair((blockposition.getX() + l) >> 4, (blockposition.getZ() + i1) >> 4);
+                        if (notGeneratedChunks.contains(chunkCoord)) continue;
+                        if (!this.world.getChunkProviderServer().isChunkGenerated(chunkCoord.x, chunkCoord.z)) {
+                            notGeneratedChunks.add(chunkCoord);
+                            continue;
+                        }
+                    }
+                    // Lamius end
                     for (BlockPosition blockposition2 = blockposition.a(l, this.world.ab() - 1 - blockposition.getY(), i1); blockposition2.getY() >= 0; blockposition2 = blockposition1) {
                         blockposition1 = blockposition2.down();
                         if (this.world.getType(blockposition2).getBlock() == Blocks.PORTAL) {
